src/Compilers/Core/Portable/RealParser.cs

```diff
@@ -0,0 +1,783 @@
+            ulong roundBitMask = (1UL << (shift - 1));
+            ulong lsbBitMask = 1UL << shift;
+
+            bool lsbBit = (value & lsbBitMask) != 0;
+            bool roundBit = (value & roundBitMask) != 0;
+            bool hasTailBits = !hasZeroTail || (value & extraBitsMask) != 0;
+
+            return (value >> shift) + (ShouldRoundUp(lsbBit: lsbBit, roundBit: roundBit, hasTailBits: hasTailBits) ? 1UL : 0);
+        }
+
+        /// <summary>
+        /// Determines whether a mantissa should be rounded up in the  
+        /// round-to-nearest-ties-to-even mode given [1] the value of the least  
+        /// significant bit of the mantissa, [2] the value of the next bit after  
+        /// the least significant bit (the "round" bit) and [3] whether any  
+        /// trailing bits after the round bit are set.  
+        ///  
+        /// The mantissa is treated as an unsigned integer magnitude.  
+        ///  
+        /// For this function, "round up" is defined as "increase the magnitude" of the  
+        /// mantissa.
+        /// </summary>
+        /// <param name="lsbBit">the least-significant bit of the representable value</param>
+        /// <param name="roundBit">the bit following the least-significant bit</param>
+        /// <param name="hasTailBits">true if there are any (less significant) bits set following roundBit</param>
+        /// <returns></returns>
+        private static bool ShouldRoundUp(
+            bool lsbBit,
+            bool roundBit,
+            bool hasTailBits)
+        {
+            // If there are insignificant set bits, we need to round to the  
+            // nearest; there are two cases:  
+            // we round up if either [1] the value is slightly greater than the midpoint  
+            // between two exactly representable values or [2] the value is exactly the  
+            // midpoint between two exactly representable values and the greater of the  
+            // two is even (this is "round-to-even").  
+            return roundBit && (hasTailBits || lsbBit);
+        }
+
+        /// <summary>
+        /// Multiply a BigInteger by the given power of two.
+        /// </summary>
+        /// <param name="number">The BigInteger to multiply by a power of two and replace with the product</param>
+        /// <param name="shift">The power of two to multiply it by</param>
+        private static void ShiftLeft(ref BigInteger number, uint shift)
+        {
+            var powerOfTwo = BigInteger.Pow(BigTwo, (int)shift);
+            number = number * powerOfTwo;
+        }
+
+        /// <summary>
+        /// Multiply a BigInteger by the given power of ten.
+        /// </summary>
+        /// <param name="number">The BigInteger to multiply by a power of ten and replace with the product</param>
+        /// <param name="power">The power of ten to multiply it by</param>
+        private static void MultiplyByPowerOfTen(ref BigInteger number, uint power)
```

Comment:
Not sure if performance is an issue here for large exponents, but can be made faster with a lookup table, but would probably be too much overhead for just a compiler.